### ISSUE TITLE
How to do cross Validation with yolov8?

### ISSUE THREAD

**Issue by srikar242** [ultralytics/ultralytics#13560]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

@glenn-jocher How to do cross validation with yolov8? I am trying to use StratifiedKFold from sklearn to do it. My Question is how to convert my yolov8 model into sklearn compatible estimator? I have make yolov8 compatible with scikit-learn's cross-validation functions I guess. Am I right? If not, please tell me the steps that I need to take to do this..

**Comment by glenn-jocher** [comment]
@srikar242 hello!

Thank you for your question. Cross-validation is a great way to ensure your model's robustness and generalizability. While YOLOv8 is not directly compatible with scikit-learn's `StratifiedKFold`, you can still perform cross-validation by manually splitting your dataset and training the model on each fold. Here's a step-by-step guide to help you achieve this:

1. **Setup**: Ensure you have the necessary libraries installed:
    ```bash
    pip install -U ultralytics scikit-learn pandas pyyaml
    ```

2. **Prepare Your Dataset**: Ensure your annotations are in the YOLO format. For this example, let's assume your dataset is structured with images and labels in separate directories.

3. **Generate Feature Vectors**: Create a feature vector for each image based on the presence of class labels. This will help in stratifying the dataset.
    ```python
    import pandas as pd
    from collections import Counter
    from pathlib import Path

    dataset_path = Path("./path/to/dataset")
    labels = sorted(dataset_path.rglob("labels/*.txt"))

    # Assuming your classes are defined in a YAML file
    with open("path/to/data.yaml", "r") as y:
        classes = yaml.safe_load(y)["names"]
    cls_idx = sorted(classes.keys())

    labels_df = pd.DataFrame([], columns=cls_idx, index=[l.stem for l in labels])

    for label in labels:
        lbl_counter = Counter()
        with open(label, "r") as lf:
            lines = lf.readlines()
        for l in lines:
            lbl_counter[int(l.split(" ")[0])] += 1
        labels_df.loc[label.stem] = lbl_counter

    labels_df = labels_df.fillna(0.0)
    ```

4. **K-Fold Split**: Use `KFold` from `sklearn` to split the dataset.
    ```python
    from sklearn.model_selection import KFold

    ksplit = 5
    kf = KFold(n_splits=ksplit, shuffle=True, random_state=20)
    kfolds = list(kf.split(labels_df))
    ```

5. **Training with Cross-Validation**: Train your YOLOv8 model on each fold.
    ```python
    from ultralytics import YOLO

    weights_path = "path/to/weights.pt"
    model = YOLO(weights_path, task="detect")

    results = {}
    batch = 16
    project = "kfold_demo"
    epochs = 100

    for k, (train_idx, val_idx) in enumerate(kfolds):
        train_files = labels_df.iloc[train_idx].index
        val_files = labels_df.iloc[val_idx].index

        # Create dataset YAML for each fold
        dataset_yaml = f"split_{k+1}_dataset.yaml"
        with open(dataset_yaml, "w") as ds_y:
            yaml.safe_dump({
                "path": dataset_path.as_posix(),
                "train": train_files.tolist(),
                "val": val_files.tolist(),
                "names": classes,
            }, ds_y)

        model.train(data=dataset_yaml, epochs=epochs, batch=batch, project=project)
        results[k] = model.metrics
    ```

This approach ensures that you manually handle the cross-validation process while leveraging the power of YOLOv8 for training and evaluation.

For a more detailed guide, you can refer to our [K-Fold Cross Validation documentation](https://docs.ultralytics.com/guides/kfold-cross-validation/).

Feel free to reach out if you have any more questions. Happy coding! 😊

**Comment by FisherYuuri** [comment]
> @srikar242 hello!
> 
> Thank you for your question. Cross-validation is a great way to ensure your model's robustness and generalizability. While YOLOv8 is not directly compatible with scikit-learn's `StratifiedKFold`, you can still perform cross-validation by manually splitting your dataset and training the model on each fold. Here's a step-by-step guide to help you achieve this:
> 
> 1. **Setup**: Ensure you have the necessary libraries installed:
>    ```shell
>    pip install -U ultralytics scikit-learn pandas pyyaml
>    ```
> 2. **Prepare Your Dataset**: Ensure your annotations are in the YOLO format. For this example, let's assume your dataset is structured with images and labels in separate directories.
> 3. **Generate Feature Vectors**: Create a feature vector for each image based on the presence of class labels. This will help in stratifying the dataset.
>    ```python
>    import pandas as pd
>    from collections import Counter
>    from pathlib import Path
>    
>    dataset_path = Path("./path/to/dataset")
>    labels = sorted(dataset_path.rglob("labels/*.txt"))
>    
>    # Assuming your classes are defined in a YAML file
>    with open("path/to/data.yaml", "r") as y:
>        classes = yaml.safe_load(y)["names"]
>    cls_idx = sorted(classes.keys())
>    
>    labels_df = pd.DataFrame([], columns=cls_idx, index=[l.stem for l in labels])
>    
>    for label in labels:
>        lbl_counter = Counter()
>        with open(label, "r") as lf:
>            lines = lf.readlines()
>        for l in lines:
>            lbl_counter[int(l.split(" ")[0])] += 1
>        labels_df.loc[label.stem] = lbl_counter
>    
>    labels_df = labels_df.fillna(0.0)
>    ```
> 4. **K-Fold Split**: Use `KFold` from `sklearn` to split the dataset.
>    ```python
>    from sklearn.model_selection import KFold
>    
>    ksplit = 5
>    kf = KFold(n_splits=ksplit, shuffle=True, random_state=20)
>    kfolds = list(kf.split(labels_df))
>    ```
> 5. **Training with Cross-Validation**: Train your YOLOv8 model on each fold.
>    ```python
>    from ultralytics import YOLO
>    
>    weights_path = "path/to/weights.pt"
>    model = YOLO(weights_path, task="detect")
>    
>    results = {}
>    batch = 16
>    project = "kfold_demo"
>    epochs = 100
>    
>    for k, (train_idx, val_idx) in enumerate(kfolds):
>        train_files = labels_df.iloc[train_idx].index
>        val_files = labels_df.iloc[val_idx].index
>    
>        # Create dataset YAML for each fold
>        dataset_yaml = f"split_{k+1}_dataset.yaml"
>        with open(dataset_yaml, "w") as ds_y:
>            yaml.safe_dump({
>                "path": dataset_path.as_posix(),
>                "train": train_files.tolist(),
>                "val": val_files.tolist(),
>                "names": classes,
>            }, ds_y)
>    
>        model.train(data=dataset_yaml, epochs=epochs, batch=batch, project=project)
>        results[k] = model.metrics
>    ```
> 
> This approach ensures that you manually handle the cross-validation process while leveraging the power of YOLOv8 for training and evaluation.
> 
> For a more detailed guide, you can refer to our [K-Fold Cross Validation documentation](https://docs.ultralytics.com/guides/kfold-cross-validation/).
> 
> Feel free to reach out if you have any more questions. Happy coding! 😊

Hello, I found that this approach leads to each round of training inheriting the training results from the previous dataset, i.e. the first epoch will get the highest validation accuracy, I'm not sure if this is correct or not

**Comment by ovidedecroly** [comment]
@glenn-jocher Hello,
If I want to create a research report or paper using StratifiedKFold, do I need testing data to create a confusion matrix and other metrics? Or just training and valid data?

**Comment by glenn-jocher** [comment]
@FisherYuuri You’re seeing that because the same `YOLO` instance (and run dir) is reused across folds—re‑init the model each fold, disable resume, and give each run a unique `name` so every fold starts from identical weights:
```python
from ultralytics import YOLO

for k, dataset_yaml in enumerate(ds_yamls, 1):
    model = YOLO("yolov8n.pt")  # re-init per fold
    model.train(data=dataset_yaml, epochs=epochs, batch=batch,
                project="kfold_demo", name=f"fold_{k}", resume=False, seed=0)
```
If anything still carries over, ensure you’re not writing into a previous `project/name` and that no auto-resume is triggered.

@ovidedecroly In K‑Fold you can report the mean/STD of validation metrics across folds, but for a paper you should keep a held‑out test split and report final metrics (and the confusion matrix) on that never‑seen test set:
```python
from ultralytics import YOLO

model = YOLO("path/to/best.pt")
metrics = model.val(data="data.yaml", split="test", plots=True, save_json=True)  # saves confusion matrix
```
See the validation options in the [Model Validation docs](https://docs.ultralytics.com/modes/val/) and a high-level recap in the [Cross‑Validation overview](https://www.ultralytics.com/glossary/cross-validation).

This issue is closed and marked stale—if you need follow‑up or hit a new problem, please open a new issue for visibility.
